Added a new feature that keeps track of known contracts

RCL-Carl opened the pull request:
When using the populus workflow, I find myself being in this position where I copy/pasta the address of the deployed contract over and over. This is particularly annoying for me because I don't use a mouse. 

This pull request has some code that stores the contract address for each deployed contract in a "known_contracts.json" file associated with the test chain it deployed to. This json file contains the address, timestamp, and the SHA-512 hash of the solidity binary code. When the user runs the attach sub-command, they can select the test chain that they want to run against, and then the code will look at the "known_contracts.json" file to figure out what instances of the deployed contracts to make available in the user environment. It compares the SHA-512 hash of the latest solidity binary code and only instantiates known contracts for which these hashes match. 

This is a WIP. The current implementation is not very user friendly. Basically, the way the code is structured, the "chain run" method is separate from the attach method. This means that the attach method does not know the details of what chain is running (this is obviously by design and mirrors the way geth works). But what this means is I need an additional piece of information when I run attach to get the known contracts of that test chain. 

I'm proposing that we add a new sub-command "console" which would take the same arguments as the "chain run" command, but would be an amalgam of the "chain run" and "attach" commands. It would launch the "chain run" command in the background and then launch the "attach" console with the data_dir from the "chain run" command to figure out the known contracts. The user would still be able to see the log of the geth miner process by using "tail -f ./chains/default/logs/geth*.log". 

![Armadillo](http://i.imgur.com/mVBDL8r.jpg)

pipermerriam commented:
> I'm proposing that we add a new sub-command "console" which would take the same arguments as the "chain run" command, but would be an amalgam of the "chain run" and "attach" commands.

What if the active chain (being run by `populus chain run`) was symlinked to a known common location such as `./chains/.active-chain/ > ./chains/default/`.  This would let `populus attach` just look in a single location.

Pros:
- eliminates the need for a new command.
- removes complexity of figuring out what to do with geth output within `populus console` since it's beneficial to be able to see what geth is doing.

Cons:
- Requires additional logic in `populus chain run` to deal with things like symlinks that were not cleaned up correctly.

Thoughts.  I'm not sure this is better, but it's at least an alternative implementation with different trade-offs.

RCL-Carl commented:
The only other con I would say for that approach (and I realize this is some what obscure) is that you would not be able to run two separate instances of populus chain run from the same project without them messing up each others reference to ".active-chain". I'm not sure that anyone would actually do this, but I can maybe see a use case where we want to parallelize test runs in the future.

pipermerriam commented:
> you would not be able to run two separate instances of populus chain run from the same project without them messing up each others reference to ".active-chain"

What about adding a new `--no-symlink` flag to `chain run`?  That could potentially eliminate the problem.

RCL-Carl commented:
Ya... I think there is another issue as well. If you consider the use case, a user starts up a chain and then starts up an attach terminal. Then lets say the user finds a bug in their contract. They are going to shutdown the chain, recompile, invoke deploy, and then rerun the chain. I can imagine a high number of people will also forget to restart attach. So now attach either needs to watch for changes with inotify and reload the known contracts or tell the user they are out of sync and to restart the attach terminal. I think reloading known contracts becomes dicey when you start considering other variables referencing the old known contracts objects, to say nothing of the actual contract class types.

I think in general a "console" command that runs the test chain and synchronizes the data is cleaner and will lead to less confusion for the user. I think we should leave attach as it is currently implemented alone so that it doesn't load any known contracts so that it can be of general purpose use on any ethereum chain that it attempts to connect to (ie, frontier, or tests chains). Are there any cons to this approach that I'm not addressing?

RCL-Carl commented:
removes complexity of figuring out what to do with geth output within populus console since it's
    beneficial to be able to see what geth is doing.

I guess I was just assuming that the geth log would continue to be logged to the chains/default/logs directory and that the user could use "tail -f" fairly easily (possibly with a symlink to the latest log to make that easier).

pipermerriam commented:
> I guess I was just assuming that the geth log would continue to be logged to the chains/default/logs directory and that the user could use "tail -f" fairly easily

I think this is fine.  I'd suggest that the text displayed to the user should potentially provide a copy/pastable version of this command so that it's easy for less technical user's to get access to the log.

pipermerriam commented:
>  If you consider the use case, a user starts up a chain and then starts up an attach terminal. Then lets say the user finds a bug in their contract. They are going to shutdown the chain, recompile, invoke deploy, and then rerun the chain

Yeah, that's pretty nasty.  To make sure that I understand, is the idea that when using `console` command, the user would fully exit the process, redeploy, and then re-enter using the `console` command?

One alternative way to solve this and keep everything under the attach command would be to adjust the  object that the deployed contracts are accessed on so that it *detects* changes to the compiled contracts, outputs appropriately informative warning in the event that it detects these changes, and provides helper methods for executing the redeploy from within the attached terminal so that the user never has to exit.

Also, I'm not trying to talk you into doing this all within the attach command.  I'm fine with either.  Just hoping that this discussion leads to a well thought out implementation.  Thanks for your work on this.

RCL-Carl commented:
Yeah, that's pretty nasty. To make sure that I understand, is the idea that when using console
    command, the user would fully exit the process, redeploy, and then re-enter using the console
    command?

Ya - that was my intended use pattern. 

    provides helper methods for executing the redeploy from within the attached terminal so that the user
    never has to exit.

Ya - I like the idea of having deployment tools in the 'attach' terminal even if we don't add any other features. If we are going to go that direction, then I think that the "known_contracts.json" is only really useful between different runs or if the test chain is being used for a demonstration of the technology - ie, these are cases where the contract is not actively changing. 

If the user has access to deployment tools during the active contract development phase, then I would imagine the user would just redeploy a new instance when the compiled code changes. I think the redeploy can also drive the sync of the latest contract ABI to the contract class type object defined on the 'contracts' local object of the attach terminal. The 'attach' console still needs to know what chain it is talking to here to update the "known_contracts.json" file as redeployments happen. 

I'm still not a huge fan of the way I have it now with calling out a chain by name or using the symlink ".activechain" method. Both have short comings from a user experience & logistics perspectives, respectively. I'll do some experiments and we can try out the different ways of doing it, but I feel like the "console" method covers both and simplifies the workflow. The ".activechain" method won't take much effort to implement so i'll put it here and then we can try it before I try adding the "console" method.

pipermerriam commented:
> I'm still not a huge fan of the way I have it now with calling out a chain by name or using the symlink ".activechain" method. Both have short comings from a user experience & logistics perspectives, respectively. I'll do some experiments and we can try out the different ways of doing it, but I feel like the "console" method covers both and simplifies the workflow. The ".activechain" method won't take much effort to implement so i'll put it here and then we can try it before I try adding the "console" method. 

Your call.  You're the closest to the code on this so choose whichever seems to have the best trade-offs.  I've maintained populus as being pre-release software with no strong promises to backwards compatibility yet so this is something that can be iterated on, or changed outright if it doesn't pan out.

RCL-Carl commented:
Ok - I've added the ".active-chains" symlink feature. I've additionally added a simple method to redeploy contracts. Currently, it doesn't dry run the contract creation so just FYI. I also added a file system observer with watchdog to automatically reload the 'known' contracts lists per contract if the user opens a new chain using "populus chain run".

I'm probably not going to implement the "console" command method because I want to move on to other things (like actually making contracts). This is good enough for now. 

One minor issue that I don't know how to get around: When the watchdog observer detects that the active-chain link has changed, the code outputs a message that indicates the new data directory that has been loaded. Unfortunately, I don't know how to make the interpreter console print out the "In[32]: " prefix again so that there is a smooth interpreter UI. It isn't a big deal, because the user can just press enter and it will show the prefix again, but it is kind of annoying.

pipermerriam commented on populus/cli/attach_cmd.py:
```diff
@@ -12,29 +14,172 @@
+            of strings indicating the names of the contracts to deploy
+    """
+
+    if type(contracts_by_name) is not list:
```
I think this would be better if it accepted any iterable.

pipermerriam commented on populus/cli/attach_cmd.py:
```diff
@@ -12,29 +14,172 @@
+    knownCts = get_known_contracts(data_dir)
+    for name, ctType in context["contracts"]:
+        if name in knownCts.keys():
+            addrList = knownCts[name]
```
I would prefer that the naming conventions for the local variables be in the format of `known_cts` to remain consistent with the rest of the codebase.

pipermerriam commented on populus/geth.py:
```diff
@@ -32,6 +35,27 @@ def get_geth_data_dir(project_dir, name):
+    """
+    bc_dir = get_blockchains_dir(project_dir)
+    active_dir = os.path.abspath(os.path.join(bc_dir, ACTIVE_CHAIN_SYMLINK))
+    return(active_dir)
```
Can you remove the parens from this statement so that it is `return active_dir`

pipermerriam commented:
Just a few comments on code style to keep the codebase consistent.  If you can get to them, I'll get this merged and released ASAP.  Otherwise, it may be a bit later in the week before I can get to the various bits of cleanup.  Thanks again.

RCL-Carl commented:
I've pushed formatting changes and I just removed the check for a list type in favor of duck typing for iterable.

pipermerriam commented:
You can get this on pypi in the 0.6.3 release.

RCL-Carl commented:
Quick test using virtualenv show that this deploys cleanly and works as intended.
